### PR TITLE
Enable inlining of a subroutine with an array subrange in an argument

### PR DESCRIPTION
**Pull request by awnawab**
A small fix to the inlining trafo that enables inlining of calls where we pass a subrange of an array argument, i.e.:
```
real :: a(8)

!$loki inline
call kernel(a(5:8)

....

subroutine kernel
real, intent(in) :: a(4)

```

### CODE REVIEW

File: loki/transformations/inline/tests/test_inline_transformation.py [hunk 1]
```diff
@@ -338,18 +339,24 @@ def test_inline_transformation_adjust_imports(frontend, tmp_path):
...
+  real(kind=8), intent(out) :: c(4), d(4)
   real(kind=8) :: tmp(m)
   integer :: i
 
   tmp(1:m) = x
   do i=1, n
     a(i) = b(i) + sum(tmp)
   end do
+  do i=1,4
+    c(i) = 0.
```
Comment (mlange05) [inline loki/transformations/inline/tests/test_inline_transformation.py:356]: Could we add an assignment `d(i) = 1.` here too, to check that this index is shifted too?

File: loki/transformations/inline/tests/test_inline_transformation.py [hunk 2]
```diff
@@ -338,18 +339,24 @@ def test_inline_transformation_adjust_imports(frontend, tmp_path):
   implicit none
   contains
 
-subroutine test_inline_inner(a, b)
+subroutine test_inline_inner(a, b, c, d)
   use BNDS_module, only: n, m
   use another_module, only: x
 
   real(kind=8), intent(inout) :: a(n), b(n)
+  real(kind=8), intent(out) :: c(4), d(4)
```
Comment (mlange05) [inline loki/transformations/inline/tests/test_inline_transformation.py:347]: What happens to `real, intent(inout) :: e(0:3)` in the two cases below? Could we add a test for this too, just to cover our bases?

File: loki/transformations/inline/procedures.py [hunk 3]
```diff
@@ -87,12 +87,29 @@ def _map_unbound_dims(var, val):
...
+
         new_dimensions = list(val.dimensions)
 
         indices = [index for index, dim in enumerate(val.dimensions) if isinstance(dim, sym.Range)]
 
         for index, dim in enumerate(var.dimensions):
-            new_dimensions[indices[index]] = dim
+            # if the argument contains an array range, we must map the bounds accordingly
+            if isinstance(val.dimensions[index], sym.Range) and (lower := val.dimensions[index].lower):
+                if isinstance(dim, sym.Range):
```
Comment (mlange05) [inline loki/transformations/inline/procedures.py:101]: [no action] Note to self, there's a similar index-shifting problem in the associate resolver. We could possibly re-use a shared utility for this if we were to externalise this? This is beyond this PR tho.

**Review by mlange05 — CHANGES_REQUESTED** [review]
Looks great and very useful. I've flagged a few corner cases, that I think we want to cover in the tests to avoid future debug pain. Otherwise great, many thanks! 🙏 (I might just steal that for the associate resolve some time. 😉 )

File: loki/transformations/inline/tests/test_inline_transformation.py (hunk 2 above)
Comment (awnawab) [reply loki/transformations/inline/tests/test_inline_transformation.py]: That was indeed a corner case I'd missed, thanks!

**Comment by awnawab** [comment]
Please wait to merge this, I think I've missed the edge case where we are passing a subrange of a dimension and there are different lbounds for the argument and dummy. I'll confirm today and push a fix + test if necessary.

**Review by mlange05 — APPROVED** [review]
Looks great! Many thanks for extending to the corner cases. GTG :shipit:
